### PR TITLE
Competition report

### PR DESCRIPTION
**Pull request by sethtroisi**
screenshot: https://i.imgur.com/SrU6Dpy.png

I've included a version using sprintf which is more compact and a version using QString.arg which is a little more QT-esque

I've left in my hack to run() which enables this to be tested without playing real games.

### CODE REVIEW

**Comment by zhurak** [comment]
Why current best network is 92c658d7 if it is clearly worse than previous one a1151640?

**Comment by gcp** [comment]
I think it is because his testing code is generating fake game results.

**Comment by zhurak** [comment]
Ah, false alarm then.

**Comment by marcocalignano** [comment]
I am sorry but for me sprintf is C and it is ugly in a C++/QT code. but it 's my opinion I'd welcome a version with QStremText Object like before.

**Comment by sethtroisi** [comment]
I like sprintf but I got the sense that you & gcp prefer QT code so I added a QT version as well (see Validation.cpp line 255)

**Comment by marcocalignano** [comment]
But the sprintf is still there you could build the string using QString (append, prepend, operator+ etc.)

**Comment by sethtroisi** [comment]
Your comment, particularly "you could", makes me thing you haven't looked at the code I wrote.

I included both versions (a [block of code that used sprintf to produce the report](https://github.com/gcp/leela-zero/pull/186/commits/3e7e3b245c9c67b3c04b9c4638d9b363bdc6d500#diff-3746371f9b46ece1340a2a556a47a633R232) and a [separate block of code that used QString to produce the report](https://github.com/gcp/leela-zero/pull/186/commits/3e7e3b245c9c67b3c04b9c4638d9b363bdc6d500#diff-3746371f9b46ece1340a2a556a47a633R255)) in the pull request.

Before this message I hadn't deleted the sprintf code (that's done now as of [3da6ff7](https://github.com/gcp/leela-zero/pull/186/commits/3da6ff76aab4c7d7da59b97961dfde1006c8d4cc)) because I don't understand how gcp prefers to do pull requests.

It's late, in the morning I'll file an issue asking for qualification on how @gcp would like the pull requests process to happen.

**Comment by marcocalignano** [comment]
Yep I stop reading when I saw the sprintf, sorry! Like I said I like more the QT style because is clearer. For you (and me using it every day) sprint is fast, clear and compact, but not for everyone else. burI guess the choice has to come from gcp.

**Comment by gcp** [comment]
I have no strong feelings about printf style formatting versus C++ style, but 
```
char report[4][100];" 
```
is code that I will avoid merging whenever possible.

I like the report but abusing the Sprt class to act also act a holder for the game results for color isn't very nice. I think the end result is going to look a lot cleaner if you make a custom class for that.

**Comment by sethtroisi** [comment]
https://www.youtube.com/watch?v=CxK_nA2iVXw

I can add a custom class for storing game results.
Do you mind if sprt continues to store it's own copy of the game results?

**Comment by sethtroisi** [comment]
I moved storing games results (and printing the results to a new file (Results.cpp),

Would appreciate another look when you get time.

File: autogtp/Validation.cpp [hunk 1]
```diff
@@ -21,30 +21,39 @@
 #include <QFile>
 
 void ValidationWorker::run() {
-     do {
+    // TODO delete before code is merged, useful for testing output.
```
[the remark this answers is not in the document]
Comment (sethtroisi) [reply autogtp/Validation.cpp]: Done

File: autogtp/Results.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    void addGameResult(Sprt::GameResult result, int side);
+    void printResults(QString firstNetName, QString secondNetName);
+
+private:
+    int m_gamesPlayed;
+    int m_blackWins;
+    int m_blackLosses;
+    int m_whiteWins;
+    int m_whiteLosses;
+	mutable QMutex m_mutex;
```
[the remark this answers is not in the document]
Comment (sethtroisi) [reply autogtp/Results.h]: I copied this pattern from from SPRT.h but I'm much happier with atomics :)

File: autogtp/Results.cpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        /*
+        Produces reports in this format.
+        leelaz-ABCD1234 v leelaz-DEFG5678 (176 games)
+                        wins          black       white
+        leelaz-ABDC1234   65 36.93%   37 42.53%   28 31.46%
+        leelaz-DEFG5678  111 63.07%   61 68.54%   50 57.47%
+                                      98 55.68%   78 44.32%
+        */
+
+        QString first_name = "leelaz-" + firstNetName.left(8);
```
[the remark this answers is not in the document]
Comment (sethtroisi) [reply autogtp/Results.cpp]: done

File: autogtp/Results.cpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+#include "Results.h"
+#include "Game.h"
+#include "SPRT.h"
+#include <QTextStream>
+#include <boost/format.hpp>
+#include <boost/algorithm/string.hpp>
+
+
+void Results::addGameResult(Sprt::GameResult result, int side)
+{
```
[the remark this answers is not in the document]
Comment (sethtroisi) [reply autogtp/Results.cpp]: done

File: autogtp/Results.cpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    You should have received a copy of the GNU General Public License
+    along with Leela Zero.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "Results.h"
+#include "Game.h"
+#include "SPRT.h"
+#include <QTextStream>
+#include <boost/format.hpp>
+#include <boost/algorithm/string.hpp>
```
[the remark this answers is not in the document]
Comment (sethtroisi) [reply autogtp/Results.cpp]: Leela already has this dependency in src/ so I assumed it was okay to use. 

I really didn't like the QTString formatting (I tried it [here](https://github.com/gcp/leela-zero/pull/186/commits/496e69a6b921be5d95f715072980641d4da12678#diff-3746371f9b46ece1340a2a556a47a633))

File: autogtp/Results.cpp [hunk 6]
```diff
@@ -0,0 +1,100 @@
+/*
+    This file is part of Leela Zero.
+    Copyright (C) 2017 Marco Calignano
```
[the remark this answers is not in the document]
Comment (sethtroisi) [reply autogtp/Results.cpp]: done

**Comment by marcocalignano** [comment]
I like it, really clean!

**Comment by gcp** [comment]
One disadvantage is that there's now a (theoretical though practically non-existent) possibility to display stats that don't add up, because reading a bunch of statistics is only atomic in the individual values, not in their aggregate. I didn't realize this when I was talking about considering making them atomics - my objection was more over the strangeness of having a "mutable" in there without anything taking it by const reference.

I need to check if I can still manage Windows builds now that I end up with a Qt and boost dependency, probably the Qt project file will need some adjustment.

**Comment by marcocalignano** [comment]
That is true. But you can solve this if you put a QMutexLocker locker(&m_mutex); in getWDL() of Sprt
and if there is also a mutex in the Result object similar to Sprt (and you can then save the atomic

**Comment by sethtroisi** [comment]
I added back mutex

**Comment by sethtroisi** [comment]
I removed the mutex because of the sync_mutex added to validation.

I'd be nice if this could go in before the next release, to help better understand the black / white imbalance while training networks.

**Comment by gcp** [comment]
I removed the boost dependency, but I'm pretty sure that didn't cause the wonky math:

```
w v w ( 39 games)
              wins        black       white
    6 15.38%    6 15.38%    0  0.00%    6 31.58%
   33 84.62%   33 84.62%   13 65.00%   20 105.26%
                       13 33.33%   26 66.67%
```

**Comment by gcp** [comment]
I think I got it. The p1_white_games/black_games for the second player were reversed.

**Pull request merged**